### ISSUE TITLE
[Feature request] Allow MAX_FRAGMENT_SIZE to take on values greater than USHRT_MAX

### ISSUE THREAD

**Issue by andrewbartolo** [DynamoRIO/dynamorio#4707]
Hi again,

I've been developing a fairly heavy-duty instrumentation client that, among other things, wraps and dumps all memory load and store instructions, à la memtrace_simple. Recently, I compiled the SPEC CPU2017-rate suite for x86-64 with gcc 10.2 at `-O3`, and saw the following on `cam4_r`, `parest_r`, and `wrf_r`:

`Basic block or trace instrumentation exceeded maximum size.  Try lowering -max_bb_instrs and/or -max_trace_bbs.`

I ran DR with `-debug -loglevel 3`, and discovered that, as expected, gcc was doing (an admittedly incredible amount of) inlining and loop unrolling to produce a basic block w/instrumented size exceeding 64KiB.

So, it seems like alternatives at this point are either to lower `-max_bb_instrs/-max_trace_bbs`, which entails missing a large and relevant portion of the workload, or to increase `MAX_FRAGMENT_SIZE` (and probably increase `-max_bb_instrs/-max_trace_bbs` after that) in order to support larger fragments.

Currently, MAX_FRAGMENT_SIZE is maxing out the size of the ushort fragment size field. I took a rudimentary stab at changing this and all related fields from ushort to uint, by just grepping through the codebase. I've attached a patch that attempts to do this, at least for just Linux: [max_fragment_size_diff.txt](https://github.com/DynamoRIO/dynamorio/files/5921258/max_fragment_size_diff.txt). There's something wrong, though, as the workloads progress well past the point where they asserted out previously, but throw a meta-instruction fault assertion later on.

On one hand, this is a pretty heavyweight use case. On the other hand, the trend of aggressive inlining probably isn't going away, and two extra bytes per fragment (?) for a larger size field seems like a small amount of overhead.

Any thoughts? And happy to help any way I can. Thanks!

**Comment by derekbruening** [comment]
> So, it seems like alternatives at this point are either to lower -max_bb_instrs/-max_trace_bbs, which entails missing a large and relevant portion of the workload,

Nothing would be missed: blocks larger than the maximum are simply split up into multiple blocks.

**Comment by derekbruening** [comment]
At one point in the past, when DR was run in 60 processes simultaneously on one production machine, DR was heavily optimized to reduce heap usage.  For the default setup, DR's heap uses more memory than its code cache, and so every structure's fields were shrunk and squished, resulting in significant savings.  For many use cases today, though, a user is running one tool on one application which has a machine all to itself, where none of that really matters, so increasing the heap size has little downside.  There are some uses cases where memory usage does matter, however, and that is the only reason we would hesitate before increasing things.  Increasing the fragment size might also affect other things like code cache management fields or linkstub offset ranges or exit stub reachability on RISC architectures, so it may not be trivial.

Before going that route, though, it might be much easier to just reduce -max_bb_instrs.  A PR to reduce the default value further seems reasonable and has little downside.  Please see my prior comment: unless I misread your initial post, it sounds like you thought reducing the maximum would have some negative effect like skipping large blocks; it will not.  There is almost zero downside to decreasing the maximum (the only downside is a smaller scope for intra-block optimizations and analyses for these large blocks).

**Comment by andrewbartolo** [comment]
Thanks a lot for the explanation + justification. I was indeed under the mistaken impression that reducing `-max_bb_instrs` would cause the basic block to be skipped entirely, rather than split into multiple smaller blocks, as actually occurs. This is a nice feature, and with it, I don't need the larger size field.

After bumping down `-max_bb_instrs`, I did run into another (unrelated) issue with `drutil_insert_get_mem_addr()` generating seemingly inaccurate memory reference addresses.

Before starting another issue in the tracker, at a high level, is there any reason why the following would cause a segfault? It happens after ~10s or so of execution, which seems odd.

In the app_instruction callback:
1. Insert a clean call to dereference a memory address stored in a TLS slot by the previous instruction (if any).
2. If the current instruction does a LOAD, get that LOAD's addr. with `drutil_insert_get_mem_addr()` and store it in the aforementioned TLS slot.

I have a minimized DR client for this, and I did also account for the drreg restore-elide issue in #2449 / #2472. There was also something mentioned about some TLS slots not being valid across instructions, but I don't think this is happening, as I'm only using a single one.

**Comment by derekbruening** [comment]
> I did run into another (unrelated) issue with drutil_insert_get_mem_addr() generating seemingly inaccurate memory reference addresses.

We would very much like to know about inaccuracies in drutil_insert_get_mem_addr().  But if you're not sure whether it's a bug it is generally best to first discuss on the users list https://groups.google.com/forum/#!forum/DynamoRIO-Users as it will reach a wider audience of people who might have an answer, and it will reach other users who may find the information beneficial.

**Comment by derekbruening** [comment]
Thanks, I see https://groups.google.com/g/dynamorio-users/c/8_kvLY4YR-M

Lowering priority of this feature issue; maybe could consider closing.
